Cadence v5 feature upgrade

za419 opened the pull request:
This PR is currently incomplete and untested, but I don't mind putting it out for comments in this state.

I also realize I probably should have broken up the scope here and merged each feature individually - Oh well.

za419 committed:
Implement config for the history command

za419 committed:
Implement history command

This isn't tested yet, but I think it should work...

I'll debug everything in one go later.

za419 committed:
Add config for auto-listenURL

I want to provide a fallback in case the fetch fails, and it's
convenient this way in case we turn off auto-listenURL for debugging
purposes.

za419 committed:
Implement auto-listenURL

za419 commented on bot.js:
```diff
@@ -117,13 +117,22 @@ const stream = bot.voice.createBroadcast();
 
+// This will be the path we connect to.
+// By default, we shall use fallback paths set in config
+// But, if configured to try to auto-fetch it, we'll grab the new path and put it here.
```
I considered also adding behavior for the fallback path to get used later if the stream fails to connect or dies - As a hardcoded redundancy, not just a block to the listenURL not getting updated.

I spent about 10 seconds on it and didn't think of a great way to differentiate a temporary network issue from a condition where we should set back to fallback, except implementing a reconnect counter and tuning it for some big number, which seems sad.

Thoughts? Would this be expected behavior I should put in?

za419 commented on bot.js:
```diff
@@ -138,6 +147,38 @@ function beginGlobalPlayback() {
     }
 }
 
+// If the stream path should be set automatically, try to override our streamPath from the upstream.
```
With this, the minimal configuration for a new CadenceBot deployment connected to a new CadenceBot server would be something like

```json
{
  "API": {
    "aria": {
      "prefix": "https://NEW_SERVER_HERE"
    }
  }
}
```

as long as default command triggers and such are comfortable for the new IA.

I don't think we'll get it any lower than that, so @kenellorando if you want to anticipate the install script generating a `config.json` like that for CadenceBot feel free to do so.

za419 commented on bot.js:
```diff
@@ -1169,22 +1208,26 @@ function command(message) {
                     );
 
                     // Grab the report of how much time is left from the response, and parse it into a string
-					try {
```
Woah, woah, woah - How the heck did *tabs* get in here!?

Goddammit... Um, me? VSCode?

Yeah, let's go with VSCode.

za419 commented on bot.js:
```diff
@@ -1327,6 +1370,52 @@ function command(message) {
+                let history;
+                if ((body = "" || (history = JSON.parse(body)).length === 0)) {
+                    log.info("Results are empty. Replying appropriately.");
+                    message.reply("Cadence has no song history at the moment.");
```
@kenellorando feel free to commend on this response and the others for `history`

kenellorando replied:
Looks fine. Cadence's UI just says `No history available (yet).`

https://github.com/kenellorando/cadence/blob/362800041145694dbf61da01833253078b670941/cadence/server/public/js/aria.js#L96-L97

za419 commented on default-config.json:
```diff
@@ -183,13 +187,17 @@
-      "prefix": "https://stream.cadenceradio.com",
-      "stream": "/cadence1"
+      "automatic": true,
+      "protocol": "https",
```
The Cadence API reference says we choose this - I decided to enforce that fallback and primary both use the same protocol, but not to enforce that that protocol matches the main API prefix.

I'm open to disagreement with that choice though.

za419 committed:
Add config for album art

za419 committed:
Take a shot at implementing album art in nowplaying

I'd like this to be a rich reply with embeds, but that requires a more
recent version of discord.js. We need to update anyway, but I'll add
that as a issue for the future rather than waiting for it here.

za419 committed:
Update package lock

za419 committed:
Don't log the whole album art

... That makes the log REALLY bad.

za419 committed:
Improve logging of path adjustment

za419 committed:
Try alternate way to reconnect

I'm not sure if this overrides the last play

za419 committed:
Add error logging for problems during auto-listenURL

za419 committed:
Add log during connection

za419 committed:
Fix mock nowplaying

za419 committed:
Try adding extension to album art attachment name

za419 committed:
Marker commit

This commit serves to record that this branch has been tested and is
ready to merge.

Therefore, this commit fixes #104.

za419 committed:
Move album art behind a memoization layer

As opposed to constantly polling, this will sharply decrease network
usage, which we quickly found to be untenable.

za419 committed:
Add a config parameter to limit album art cache

As a control on memory usage - Caching lots of album art for different
songs could potentially burn a lot of RAM.

I intend setting this to 0 to mean infinite cache size, not to disable
caching entirely - Caching at least one song should be mandatory because
to do less would potentially be very damaging to network usage.

za419 committed:
Add cache size limit and maintenance behavior
